### DISCUSSION TITLE
Use destructors instead of atexit() in drivers

### DISCUSSION THREAD

**Discussion by vg-sc** [labgrid-project/labgrid#1180] [Q&A] [answered]
I want to run tests under QEMU on multiple embedded platforms on which I'm working on using labgrid and pytest. Unfortunately, I encounter problems due to ill-timed drivers destruction. Ideally, I would like pytest to start one QEMU instance, run all tests on it, kill it, start new QEMU for the next platform and so on. Pytest tries to do that when fixture is not needed any more, but QEMUDriver doesn't actually get destroyed, because it doesn't have a destructor method. All resources remain acquired. So I end up with a dozen of useless QEMUs running simultaneously and consuming lots of memory and CPU time.

Since drivers acquire resources in their constructors, destructor seems to be the best and correct place for releasing these resources. Could you comment why labgrid instead uses atexit()?

**Comment by Emantor** [comment, marked as answer]
What you want to do is call `target.deactivate()` after you are done with qemu, this will turn off the qemu instances.

**Reply by vg-sc** [reply]
Thanks, this works. But I'm still confused about this design decision. It's very easy to leak resources, while they could be managed automatically.

**Reply by Emantor** [reply]
No, they can't. We don't know whether you want to reuse this instance of the `QemuDriver` at any point. The driver also binds to the target and remains as a reference there, so even if your local reference to the class instance is destroyed, python won't cleanup the instance since the target has a reference. In case you delete your reference using `del`, this might lead to an inconsistent target state, since afaics we don't handle this case in the target class.

**Reply by jluebbe** [reply]
If you keep using the same Target (as with the [target fixture](https://github.com/labgrid-project/labgrid/blob/master/labgrid/pytestplugin/fixtures.py#L119)), the same QEMUDriver driver is used for all tests and should manage the lifecycle of QEMU correctly.
